sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1130 @@
+
+            - True or RequestType.REQUESTED: {metadata} is requested, and passed to \
+{method} if provided.
+
+            - False or RequestType.UNREQUESTED: {metadata} is not requested and the \
+meta-estimator will not pass it to {method}.
+
+            - None or RequestType.ERROR_IF_PASSED: {metadata} is not requested, and \
+the meta-estimator will raise an error if the user provides {metadata}
+
+            - str: {metadata} should be passed to the meta-estimator with this given \
+alias instead of the original name.
+
+            The default (UNCHANGED) retains the existing request. This allows
+            you to change the request for some parameters and not others.
+
+"""
+REQUESTER_DOC_RETURN = """        Returns
+        -------
+        self : estimator
+            Returns the object itself.
+"""
+
+
+class MethodMetadataRequest:
+    """Contains the metadata request info for a single method.
+
+    Refer to :class:`MetadataRequest` for how this class is used.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    owner : str
+        The name of the object owning these requests.
+
+    method : str
+        The name of the method to which these requests belong.
+    """
+
+    def __init__(self, owner, method):
+        self._requests = dict()
+        self.owner = owner
+        self.method = method
+
+    @property
+    def requests(self):
+        return self._requests
+
+    def add_request(
+        self,
+        *,
+        prop,
+        alias,
+    ):
+        """Add request info for a prop.
+
+        Parameters
+        ----------
+        prop : str
+            The property for which a request is set.
+
+        alias : str, RequestType, or {True, False, None}
+            The alias which is routed to `prop`
+
+            - str: the name which should be used as an alias when a meta-estimator
+              routes the metadata.
+
+            - True or RequestType.REQUESTED: requested
+
+            - False or RequestType.UNREQUESTED: not requested
+
+            - None or RequestType.ERROR_IF_PASSED: error if passed
+        """
+        try:
+            alias = RequestType(alias)
+        except ValueError:
+            if not (isinstance(alias, str) and alias.isidentifier()):
+                raise ValueError(
+                    "alias should be either a valid identifier or one of "
+                    "{None, True, False}, or a RequestType."
+                )
+
+        if alias == prop:
+            alias = RequestType.REQUESTED
+
+        if alias == RequestType.UNUSED and prop in self._requests:
+            del self._requests[prop]
+        else:
+            self._requests[prop] = alias
+
+        return self
+
+    def _get_param_names(self, original_names):
+        """Get the names of all available metadata.
+
+        This method returns the names of all metadata, even the UNREQUESTED
+        ones.
+
+        Parameters
+        ----------
+        original_names : bool
```

Comment:
This name seems a bit ambiguous. "original" may be the original name given to the data (i.e. the alias passed in `fit`), or the name expected for consmption.